core/src/main/scala/chisel3/experimental/Probe.scala

```diff
@@ -0,0 +1,88 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package chisel3.experimental
+
+import chisel3._
+import chisel3.Data.ProbeInfo
+import chisel3.internal.Builder
+import chisel3.internal.Builder.pushCommand
+import chisel3.internal.firrtl._
+
+// import firrtl.ir._
+
+/** Utilities for creating and working with probe types.
+  */
+object Probe {
+
+  /** Create a probe type from a Chisel type. This is not used for
+    * hardware.
+    */
+  private def apply_impl[T <: Data](source: => T, writable: Boolean): T = {
+    val prevId = Builder.idGen.value
+    val data = source // should only evaluate source once
+    requireIsChiselType(data)
+    val ret = if (!data.mustClone(prevId)) data else data.cloneType.asInstanceOf[T]
+    ret.probeInfo = ProbeInfo(true, writable)
+    ret
+  }
+
+  def apply[T <: Data](source: => T): T = apply_impl(source, false)
+
+  def writable[T <: Data](source: => T): T = apply_impl(source, true)
+
+  def requireIsProbe(probeExpr: Data): Unit = {
+    require(probeExpr.probeInfo.isProbe, s"expected $probeExpr to be a probe.")
+  }
```

Comment:
This may be a candidate for `private[chisel3]` or `private[this]` (whatever the file private modifier is) unless we have a good reason to make this public. Alternatively, this type of functionality would be better exposed via `DataMirror` for consistency.